Simplify react lib design

Aysnine opened the pull request:
### Description

- Simple demo that use id selector
- Advanced demo with custom dynamic components

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing (`yarn test`)
- [x] Extended the README / documentation, if necessary

Aysnine committed:
Simplify react lib design

Aysnine commented:
```tsx
function App() {
  const [uncontrolledInitialOptions] = useState<GridStackOptions>({
    // ...
    children: [
      { id: "item1", h: 2, w: 2, x: 0, y: 0 },
      { id: "item2", h: 2, w: 2, x: 2, y: 0 },
    ],
  });

  return (
    <GridStackProvider initialOptions={uncontrolledInitialOptions}>
      <Toolbar />

      <GridStackRender>
        <GridStackItem id="item1">
          <div>hello</div>
        </GridStackItem>

        <GridStackItem id="item2">
          <div>grid</div>
        </GridStackItem>
      </GridStackRender>
    </GridStackProvider>
  );
}
```

Aysnine committed:
Reorganize examples

Aysnine committed:
Update react docs

adumesny commented on react/lib/grid-stack-context.ts:
```diff
@@ -1,35 +0,0 @@
-import type { GridStack, GridStackOptions, GridStackWidget } from "gridstack";
```
hey don't delete those files. I actually wanted to start shipping them like I do for /angular 
I started looking at the code and was planning to do this weekend... I wanted to match better what I did for angular wrapper (extending WidgetItem to have similar to 'selector' + 'input' which are angular centric but conceptually the same as name + props you have.

adumesny commented on react/src/lib/grid-stack-context.ts:
```diff
@@ -0,0 +1,33 @@
+import type {
```
oh I see you moved code here... quite different. will have to take a look so we can make this maybe official soon and publish as NPM.

Aysnine replied:
Maybe we need to agree on the location of the libs, I didn't check how the wrappers for the other frameworks were packaged, so I temporarily moved the folders to make it easier for me to copy the complete code between multiple demo codebases

adumesny replied:
exact location doesn't matter. I just need to publish those as part of the NPM package so users can use the classes directly and get version fixes... angular happens to use angular\projects\lib\src\lib but what you have is fine.

we need to try and match angular wrapper syntax and simplicity whiel fully supporting nested and custom component inside grid items (gs cna handle the containers) and allow darg&drop between grid without re-creating the expensive content (you app components). that's what I did for Angular.

thank you for creating React version as I know it.

Aysnine replied:
Got it.

I haven't specifically focused on the darg&drop between grid issue, but I find it works well in this pr demo.

https://gridstack-react.pages.dev/

adumesny commented:
if you look at the [angular version](https://github.com/gridstack/gridstack.js/tree/master/angular), this is what I have
```html
<gridstack [options]="gridOptions"></gridstack>
```
Is there a reason React can't match exactly ? the above is very simple DOM and provides the FULL power of GS as it doesn't try to create DOM children directly (and sync states between 2 frameworks) ?

in the DOM 'simple way (not recommended) that closer to your simple case above I also support
```html
<gridstack [options]="gridOptions" (changeCB)="onChange($event)">
   @for (n of items; track n.id) {
    <gridstack-item [options]="n">Item {{n.id}}</gridstack-item>
  }
</gridstack>
```

so I have <gridstack> and <gridstack-item> components.

Aysnine commented:
I don't have experience with angular so I didn't check its example. But it seems necessary to learn angular examples, I'll go back and confirm.

Aysnine commented:
I'm gonna set up this pr as a draft. We can align the lib design here.

Aysnine committed:
Add the GridStackContainer and update docs

Aysnine committed:
Rename file ext name

adumesny commented:
| Toolbar, which doesn't want to be in the gridstack container dom but needs to access the gridstack instance

why ? the gristack init params says if it accepts external widgets. then STATIC `GridStack.setupDragIn()` is used to init toolbar, but it doesn't need to access the instance (there could be multiple). see two.html

Aysnine commented:
> why ? the gristack init params says if it accepts external widgets (add enter/drop events). then STATIC GridStack.setupDragIn() is used to init toolbar (adds drag events), but it doesn't need to access the instance (there could be multiple). see two.html

My example here may not be appropriate. It's more like actions like save / restore require access to instance.

The toobar I have here is not quite like the one in two.html, it's more like it's made purely to give an example of how to access the gridstack through the react context. the toolbar in two.html I haven't tried yet, it's the one that's missing from this pr, I'll follow up on that.

> why having to maintain external state ? gridstack dom elements already have a gridstack pointer (and item GridItemHTMLElement have gridstackNode) and in the angular case I also store the Angular component pointers _gridComp | _gridItemComp. so from Class or DOM I can access either one. When DOM get destroyed, those are also gone - see ngOnDestroy

As I follow the component-first principle (instead of dom-first), the GridStackProvider doesn't just store the state, but also takes care of handling its lifecycle. And it provides a more standard react context way for other components to access the gridstack instance instead of getting it from the dom.

> in the angular case I use addRemoveCB (see gsCreateNgComponents) rather than the lower renderCB because angular component (your app children) can only live inside other angular components, so <gridstack> and <gridstack-item> are both angular components, but that's an angular restrictions and I'm guessing React can be hosted inside any plain DOM. So yeah I added renderCB exactly for that case...

You've reminded me that I haven't focused on ways other than renderCB, but renderCB is enough for me at the moment. I'll go back and check.

> you can at the very least use the same dom name and [options] so the code looks mostly the same.

LOL. That's what I'm struggling with, I'm finding places in the code where it's <gridstack-xxx>, which is a word, and places where it's type GridStack. In some places it's type GridStack, two words. Since it's common in react to use big humps for components, and since word spelling is easy to recognize, I followed the name type GridStack.

> We also need to talk about how custom comp are in the JSON. the old content string was legacy stuff. In angular I subclasses and added 2 params selector + input (very angular term, the second matching component @Input vs @Output fields for params). Thinking I may want to make content be a custom type (liek generic 'data' typically is) that default to string for legacy but can be our JSON app framework field of choices or something...

I think selector + input that's a good idea.

Also, in this pr, I removed the react implementation of JSON content from the lib to make it more pure. Putting the removed code implementation into advanced is an example of this.

If selector + input is implemented in gridstack, I just need to update my grid-stack-item component implementation for quick adaptation.

Aysnine commented:
> and THANK YOU for working with me on a native React wrapper as I always wanted to get this out but don't use/know React

gridstack provides very good support for custom panels for my in-house projects, and it's worth providing a better-working react wrapper for this purpose :)

adumesny commented:
| It's more like actions like save / restore require access to instance

`GridStack.init()` or `GridStack.initAll()` return instances, and angualr wrapper `GridstackComponent` stores it when doing `this._grid = GridStack.init()` so no need to keep a global list (one fewer thing to get out of sync). it's up to the app if they want to track `<gridstack>` which has an angular way of doing.

Aysnine commented:
> If selector + input is implemented in gridstack

My expression was inaccurate, I meant to say that gridstack defines the option data type well.

> maybe make existing 'content'

What about considering new fields? Like `data` or `meta`? That way you don't have to break the `content`.

Aysnine commented:
@adumesny  I just realized a serious bug where <GridStackProvider /> assigns a value to GridStack.renderCB, and when there are multiple <GridStackProvider />'s, this causes GridStack.renderCB to be duplicated, resulting in a duplication of the id maintained in each Provider. element map maintained in each Provider is inaccurate.

Also, the id doesn't seem to be required for each item, which I hadn't considered.

adumesny commented:
internal ids have to be globaly unique (why it's a running static counter) as you can move a widget from one grid to another... user ids should be unique at least per grid so save/load can map to correct instance of a widget to update.

Aysnine commented:
> internal ids have to be globaly unique (why it's a running static counter) as you can move a widget from one grid to another... user ids should be unique at least per grid so save/load can map to correct instance of a widget to update.

Resolved.

Aysnine commented:
Another confusion:

```ts
const instance = GridStack.init(optionsRef.current, containerRef.current);
instance.on("dropped", (e) => {
  console.log("dropped", e);
  // ! subGrid not working
});
```

All event not receive about subGrid.

I'm stuck here now because I want to listen for grid changes and dynamically render the react component in the react advanced example, which is a more common scenario.

Aysnine committed:
Fix unique id and Add drag in item

Aysnine committed:
Update react readme

Aysnine committed:
Fix prepareDragDrop for GridStackHandleReInitializer

adumesny commented:
> All event not receive about subGrid.|

correct, subgrids do not currently propagate events to the top one (where you add the event handler). see #2671, which I could look into

> I'm stuck here now because I want to listen for grid changes and dynamically render the react component in the react advanced example, which is a more common scenario.

when you move an item from one grid to another (child or not) you should NOT have to re-render the component (as creating them can be expensive, making API calls all over, etc... in the anular wrapper and let GS re-parent things and everything is OK BECAUSE I don't have Angualr create the dom fomr state, but let GS drive the state and callback to wrapper to create component when needed.

Are you not doing that ? that's a must change IMO.

Aysnine commented:
| when you move an item from one grid to another (child or not) you should NOT have to re-render the component (as creating them can be expensive, making API calls all over, etc... in the anular wrapper and let GS re-parent things and everything is OK BECAUSE I don't have Angualr create the dom fomr state, but let GS drive the state and callback to wrapper to create component when needed.

That's actually what I was thinking, but it's different with dragIn (clone), which always requires the creation of a new component.

And dragIn to subGrid can't be listened to. So I implemented the rendering of the dragIn copied component in a different way, instead of relying on listening. But the downside is that it's not good for save/load.

adumesny commented:
| it's different with dragIn (clone), which always requires the creation of a new component.

right, drag froms idebar is always  apreview and actual comp needs to be created. instead it's cleaner, and what I have in angualr wrapper, to use the WidgetItem as what to create and use `GridStack.renderCB` or `GridStack.addRemoveCB` which is also used during reading, to create the right component bu storing custom info on what to create (angular 'selector' field). React should do the same.

I really want React and Angular wrapper doc to look and act the same. so again I suggest you study what I do and do the equicalent react code version if not totally weird for that framework. GS needs to maintain state and re-parent and resize things as it needs to, independ on the framework. the framework only job is to create the right element (or `<gridstack>` <gridstack-item>` parent as angualr needs entire hierachy to be ng components) instead of plain divs with classes.

Aysnine commented:
> I really want React and Angular wrapper doc to look and act the same.

Yes. I want to run /angular folder. But error when install deps:

```
➜ gridstack.js/angular master ✓ yarn
yarn install v1.22.22
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
error Error: https://usw1.packages.broadcom.com/artifactory/api/npm/tis-npm-virtual/@angular/core/-/core-14.3.0.tgz: Request failed "403 Forbidden"
    at ResponseError.ExtendableBuiltin (/Users/jace/.cache/node/corepack/v1/yarn/1.22.22/lib/cli.js:696:66)
    at new ResponseError (/Users/jace/.cache/node/corepack/v1/yarn/1.22.22/lib/cli.js:802:124)
    at Request.<anonymous> (/Users/jace/.cache/node/corepack/v1/yarn/1.22.22/lib/cli.js:66750:16)
    at Request.emit (node:events:524:28)
    at module.exports.Request.onRequestResponse (/Users/jace/.cache/node/corepack/v1/yarn/1.22.22/lib/cli.js:142287:10)
    at ClientRequest.emit (node:events:524:28)
    at HTTPParser.parserOnIncomingClient (node:_http_client:716:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:117:17)
    at TLSSocket.socketOnData (node:_http_client:558:22)
    at TLSSocket.emit (node:events:524:28)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Is this a package from a private registry?

adumesny commented:
| error Error: https://usw1.packages.broadcom.com/artifactory/api/npm/tis-npm-virtual/@angular/core/-/core-14.3.0.tgz

oops, sorry. let me fix that asap. yes, private repo...

adumesny commented:
sorry about private registry! fixed in #2977

Aysnine commented:
@adumesny  I have two questions about angular wrapper:

1. Is comp1 re-render when move item from grid a to grid b ?
2. What will be happen when move item from grid a to grid b and destroy grid a?

![image](https://github.com/user-attachments/assets/0ddfa09d-ab19-4457-a0e3-9816854ec4f3)

I'm wondering if the comp1 lifecycle goes with the grid, or with the dom ref, or something else.

adumesny commented:
1. no delete/re-create. you can see that using the /angular demo (yarn + yarn start), pick two grid demo, and look at the log while dragging an object B to the other grid
![image](https://github.com/user-attachments/assets/111f6b24-8960-479b-ad0d-6e4c54aff365)

Aysnine commented:
> angular can have dynamic list of children and I'm pretty sure when GS manually reparents a child the parent 'container' doesn't try to destroy an old child...

I think this may be the essence of why react wrappers are not easy to design. react's rendering style dictates that item content cannot be designed as a react component that can be re-parented.

Same two questions but in react:

1. no delete/re-create.
2. item content (not item container by GS) will be destroy

That's why, I've seen previous GS react wrappers or examples attempting controlled components, not uncontrolled ones like mine. It's because they want to dispatch the react component with the help of the item meta in the options state to make sure the component looks re-parent.

I need to calm down a bit. 

While it occurred to me that I could use react's multiple createRoots to get each item content component off of its parent, this is not very performant and goes against the principle of a single root for react.

adumesny commented:
ok, so how do you want to proceed ? in the end when a widget is moved from a grid to another it is best to not recreate the expensive content (could be long API calls) and let GS just do a DOM re-parenting. this doesn't break Angular wrapper as GridItem comp (which then contains actual user comp) are created dynamically and will update parents list of children for destroy.

if that's not possible in React then so be it.

chungwong commented:
I gave the new design a spin but I didn't get far. Is it ready for testing?
The first error I am getting is [getContainerByWidgetId is unimplemented yet](https://github.com/Aysnine/gridstack-react/blob/main/src/lib/grid-stack-render-context.ts#L10)

AO19 commented:
@Aysnine what is your status on this? Did you get more progress?

alekseyBatuhtin commented:
@Aysnine how is it going? Mb, do you need help?
